modules/avatar/identicon/identicon.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+// Copied and modified from https://github.com/issue9/identicon/ (MIT License)
+// Generate pseudo-random avatars by IP, E-mail, etc.
+
+package identicon
+
+import (
+	"crypto/md5"
+	"fmt"
+	"image"
+	"image/color"
+)
+
+const (
+	minImageSize = 16
+)
+
+// Identicon is used to generate pseudo-random avatars
+type Identicon struct {
+	foreColors []color.Color
+	backColor  color.Color
+	size       int
+	rect       image.Rectangle
+}
+
+// New returns an Identicon
+// size image size
+// back background color
+// fore all possible foreground colors. only one foreground color will be picked randomly for one image
+func New(size int, back color.Color, fore ...color.Color) (*Identicon, error) {
+	if len(fore) == 0 {
+		return nil, fmt.Errorf("foreground is not set")
+	}
+
+	if size < minImageSize {
+		return nil, fmt.Errorf("size %d is smaller than min size %d", size, minImageSize)
+	}
+
+	return &Identicon{
+		foreColors: fore,
+		backColor:  back,
+		size:       size,
+		rect:       image.Rect(0, 0, size, size),
+	}, nil
+}
+
+// Make generates an avatar by data
+func (i *Identicon) Make(data []byte) image.Image {
+	h := md5.New()
```

Comment:
General note, while md5 is fine to use in a non-cryptographic state - it's generally slower than most non-cryptographic hash.

Reply:
I just saw:
```
// Generate pseudo-random avatars by IP, E-mail, etc
```
If that's the case, we shouldn't be using MD5. As it's currently considered a weak cryptographic function.

Reply:
Here is a non-crypto scene, so I don't worry about it.

It's also fine to use sha256 instead, will update later.